### ISSUE TITLE
Panic (possibly) involving zsched - dump available

### ISSUE THREAD

**Issue by themisanthrope** [joyent/illumos-joyent#127]
I initially posted in your IRC channel, and was advised to open an issue here.

We have two almost identical machines (_difference being that one has SSDs for ZIL and ARC_ as it's running a lot of KVM zones). One is running SmartOS standalone, the other is running as a Triton computenode.

Both of these machines have died in the last week (one hung, one rebooted), but only one managed to create a dump. This dump is from the machine running as a Triton computenode. It is also the machine with the SSDs installed.

Having had a quick look through the dump, it would _appear_ that `zsched` is related to the issue;
```
> ::msgbuf
panic[cpu11]/thread=fffffea394e578c0: 
BAD TRAP: type=e (#pf Page fault) rp=fffffe426ac79bc0 addr=90 occurred in module "genunix" due to a NULL pointer dereference
```
```
> ::cpuinfo
 11 fffffffffbc3d8c0  1b    0    0   1   no    no t-3    fffffea394e578c0 zsched
```

Incidentally, we were having problems with an LX Branded zone just before the crash, which looked like `ulimit` had been hit;
```
[root@538055d6-8d8f-c2aa-b201-a94c940847ee /]# ping packages.graylog2.org
-bash: fork: retry: No child processes
-bash: fork: retry: No child processes
^C-bash: fork: retry: No child processes
-bash: fork: retry: No child processes
-bash: fork: Resource temporarily unavailable
```

Attempting to shutdown the zone timed out, so I tried with `-f`. The system hung shortly after this. I did leave it alone for 30-ish minutes, hoping it'd either recover from a race, or reboot. I ended up power-cycling the box over IPMI.

After coming back online, I increased the `max_lwps` for that zone to `2000`, as it was set to `200` initially - probably a typo in initial configuration.
This, of course, could all be unrelated.

To get the dump and associated files;
```
cd /var/tmp && \
curl http://files.herpaderp.lol/joyent/github-illumos-joyent-issue-127.tar.gz.enc | \
openssl enc -aes-256-cbc -k !!!PASSWORD-HERE!!! -d | \
gtar -xz
```

**Due to this being a production system and me not knowing exactly what information is in this dump, the file is encrypted.** Please contact me via IRC (user _arg_) or via email (_social_ at _themisanthrope.co.uk_) for the password.

The files will be left in `/var/tmp/dump/` - the tools to decrypt are built into the SmartOS platform.

This machine's platform;
```
[root@bri-triw-001 (bridgehead) ~]# uname -a
SunOS bri-triw-001 5.11 joyent_20160707T032940Z i86pc i386 i86pc
```

The other box (that rebooted) unfortunately didn't create any dump, but it did come back online of it's own accord. We didn't lose any data due to either crash. To my knowledge, nobody was connected to the GZ at the time. This is a standalone box;
```
[root@bri-sdc-temp ~]# uname -a
SunOS bri-sdc-temp 5.11 joyent_20160707T032945Z i86pc i386 i86pc
```

Please let me know if I can provide any further information. Access to either box is an option too, if you need it.

**Comment by pfmooney** [comment]
Open as [OS-5598](https://smartos.org/bugview/OS-5598)

**Comment by h0lzi** [comment]
Did some experiemtns which nagios. I'm able to get the defunct nagios process also.
Problem here seem to be the number of workers:

/etc/nagios/nagios.cfg
# Normally worker count is dynamically allocated based on 1.5 * number of cpu's
# with a minimum of 4 workers.  This value will override the defaults
#check_workers=3

I'm able to start 6 workers. All after that are defunct.
Having more then 6 workers is not hard if you don't tune the setting.

I did not run into the fork, zone shutdown crash yet. But maybe that needs a second trigger.

**Comment by h0lzi** [comment]
little strace here:

> 
Process 37958 attached
Process 37959 attached
Process 37960 attached
Process 37961 attached
Process 37962 attached
Process 37963 attached
Process 37964 attached
Process 37965 attached
Process 37966 attached
Process 37967 attached
[pid 37958] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33) = 0
[pid 37959] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37958] write(3, "@wproc register name=Core Worker"..., 49 <unfinished ...>
[pid 37959] <... connect resumed> )     = 0
[pid 37958] <... write resumed> )       = 49
[pid 37959] write(3, "@wproc register name=Core Worker"..., 49) = 49
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37958;pid=37958) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(7, "OK\0", 3)         = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37959;pid=37959) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37960] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33) = 0
[pid 37961] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33) = 0
[pid 37960] write(3, "@wproc register name=Core Worker"..., 49) = 49
[pid 37961] write(3, "@wproc register name=Core Worker"..., 49) = 49
[pid 37957] write(8, "OK\0", 3)         = 3
[pid 37964] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37967] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37964] <... connect resumed> )     = 0
[pid 37967] <... connect resumed> )     = 0
[pid 37965] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37964] write(3, "@wproc register name=Core Worker"..., 49 <unfinished ...>
[pid 37965] <... connect resumed> )     = 0
[pid 37964] <... write resumed> )       = 49
[pid 37966] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37967] write(3, "@wproc register name=Core Worker"..., 49 <unfinished ...>
[pid 37966] <... connect resumed> )     = 0
[pid 37967] <... write resumed> )       = 49
[pid 37965] write(3, "@wproc register name=Core Worker"..., 49) = 49
[pid 37966] write(3, "@wproc register name=Core Worker"..., 49) = 49
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37962] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37960;pid=37960 <unfinished ...>
[pid 37962] <... connect resumed> )     = 0
[pid 37957] <... write resumed> )       = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37963] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37962] write(3, "@wproc register name=Core Worker"..., 49 <unfinished ...>
[pid 37963] <... connect resumed> )     = -1 ECONNREFUSED (Connection refused)
[pid 37962] <... write resumed> )       = 49
[pid 37957] write(9, "OK\0", 3)         = 3
[pid 37963] write(1, "Failed to connect to query socke"..., 106Failed to connect to query socket '/var/spool/nagios/cmd/nagios.qh': connect() failed: Connection refused
) = 106
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=37963, si_status=1, si_utime=1, si_stime=4} ---
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37961;pid=37961) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(10, "OK\0", 3)        = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37964;pid=37964) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(11, "OK\0", 3)        = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37967;pid=37967) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(12, "OK\0", 3)        = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37965;pid=37965) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(13, "OK\0", 3)        = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37966;pid=37966) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(14, "OK\0", 3)        = 3
[pid 37957] write(4, "[1472216236] wproc: Registry req"..., 71) = 71
[pid 37957] write(1, "wproc: Registry request: name=Co"..., 57wproc: Registry request: name=Core Worker 37962;pid=37962) = 57
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(15, "OK\0", 3)        = 3
[pid 37957] write(16, "################################"..., 12643) = 12643
[pid 37957] write(16, "################################"..., 13454) = 13454
Process 37968 attached
[pid 37957] write(4, "[1472216238] Successfully launch"..., 70) = 70
[pid 37957] write(1, "Successfully launched command fi"..., 56Successfully launched command file worker with pid 37968) = 56
[pid 37957] write(1, "\n", 1
)           = 1
[pid 37957] write(7, "job_id=0\0type=0\0command=/usr/lib"..., 123) = 123
Process 37969 attached
Process 37970 attached
[pid 37970] connect(4, {sa_family=AF_INET, sin_port=htons(1025), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
[pid 37970] write(1, "PING 127.0.0.1 (127.0.0.1) 56(84"..., 107) = 107
[pid 37970] write(1, "64 bytes from 127.0.0.1: icmp_se"..., 58) = 58
[pid 37970] write(1, "64 bytes from 127.0.0.1: icmp_se"..., 58) = 58
[pid 37970] write(1, "64 bytes from 127.0.0.1: icmp_se"..., 58) = 58
[pid 37970] write(1, "64 bytes from 127.0.0.1: icmp_se"..., 58) = 58
[pid 37970] write(1, "\n", 1)           = 1
[pid 37970] write(1, "--- 127.0.0.1 ping statistics --"..., 147) = 147
[pid 37969] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=37970, si_status=0, si_utime=2, si_stime=6} ---
[pid 37969] write(1, "PING OK - Packet loss = 0%, RTA "..., 105) = 105
[pid 37958] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=37969, si_status=0, si_utime=4, si_stime=13} ---
[pid 37958] write(3, "job_id=0\0type=0\0command=/usr/lib"..., 420) = 420
[pid 37957] write(4, "[1472216243] HOST ALERT: localho"..., 87) = 87
[pid 37957] write(4, "[1472216243] HOST NOTIFICATION: "..., 120) = 120
[pid 37957] write(8, "job_id=0\0type=1\0command=/usr/bin"..., 335) = 335
Process 37971 attached
Process 37972 attached
Process 37973 attached
[pid 37972] write(1, "***** Nagios *****\n\nNotification"..., 182) = 182
[pid 37973] write(3, "***** Nagios *****\n\nNotification"..., 182) = 182
[pid 37973] write(4, "Date: Fri, 26 Aug 2016 12:57:23 "..., 426) = 426
Process 37974 attached
[pid 37959] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=37971, si_status=0, si_utime=12, si_stime=42} ---
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
[pid 37974] connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("10.0.0.9")}, 16) = 0
[pid 37974] connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("10.0.0.9")}, 16) = 0
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/dev/log"}, 110) = 0
[pid 37974] write(2, "send-mail: Cannot open mail:25\n", 31) = 31
[pid 37974] connect(3, {sa_family=AF_LOCAL, sun_path="/dev/log"}, 110) = 0
[pid 37974] write(3, "\n\n***** Nagios *****\n\nNotificati"..., 184) = 184
[pid 37959] write(3, "job_id=0\0type=1\0command=/usr/bin"..., 558) = 558
[pid 37957] write(18, "################################"..., 13461) = 13461
^CProcess 37957 detached
Process 37958 detached
Process 37959 detached
Process 37960 detached
Process 37961 detached
Process 37962 detached
Process 37963 detached
Process 37964 detached
Process 37965 detached
Process 37966 detached
Process 37967 detached
Process 37968 detached
Process 37969 detached
Process 37970 detached
Process 37971 detached
Process 37972 detached
Process 37973 detached
Process 37974 detached
>

PID 37963 defuncts it tries to connect to the socket end erros:

> 
[pid 37963] connect(3, {sa_family=AF_LOCAL, sun_path="/var/spool/nagios/cmd/nagios.qh"}, 33 <unfinished ...>
[pid 37963] <... connect resumed> )     = -1 ECONNREFUSED (Connection refused)
[pid 37963] write(1, "Failed to connect to query socke"..., 106Failed to connect to query socket '/var/spool/nagios/cmd/nagios.qh': connect() failed: Connection refused
) = 106
>

this could all be unrelated. But why suddenly the socket refuses connections is still interesting.
Could  also just be timing. As with strace I can sometimes fork all workers without a problem.

**Comment by h0lzi** [comment]
The nagios problem mentioned by me is not related to OS-5598. Will create a new issue for that.

**Comment by pfmooney** [comment]
This was addressed in f7c951a817f0a182d3324e13939cd9f1edf0d9d7.
